azurerm/internal/services/keyvault/key_vault_managed_hardware_security_module_resource_test.go

```diff
@@ -0,0 +1,181 @@
+	r := KeyVaultManagedHardwareSecurityModuleResource{}
+
+	data.ResourceTest(t, r, []resource.TestStep{
+		{
+			Config: r.basic(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
+func TestAccKeyVaultManagedHardwareSecurityModule_requiresImport(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_key_vault_managed_hardware_security_module", "test")
+	r := KeyVaultManagedHardwareSecurityModuleResource{}
+
+	data.ResourceTest(t, r, []resource.TestStep{
+		{
+			Config: r.basic(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.RequiresImportErrorStep(r.requiresImport),
+	})
+}
+
+func TestAccKeyVaultManagedHardwareSecurityModule_purgeProtectionAndSoftDelete(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_key_vault_managed_hardware_security_module", "test")
+	r := KeyVaultManagedHardwareSecurityModuleResource{}
+
+	data.ResourceTest(t, r, []resource.TestStep{
+		{
+			Config: r.purgeProtectionAndSoftDelete(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
+func TestAccKeyVaultManagedHardwareSecurityModule_updateTags(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_key_vault_managed_hardware_security_module", "test")
+	r := KeyVaultManagedHardwareSecurityModuleResource{}
+
+	data.ResourceTest(t, r, []resource.TestStep{
+		{
+			Config: r.basic(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+		{
+			Config: r.updateTags(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
+func (KeyVaultManagedHardwareSecurityModuleResource) Exists(ctx context.Context, clients *clients.Client, state *terraform.InstanceState) (*bool, error) {
+	id, err := parse.HardwareSecurityModuleID(state.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := clients.KeyVault.ManagedHsmClient.Get(ctx, id.ResourceGroup, id.ManagedHSMName)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving %s: %+v", id, err)
+	}
+
+	return utils.Bool(resp.Properties != nil), nil
+}
+
+func (r KeyVaultManagedHardwareSecurityModuleResource) basic(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_key_vault_managed_hardware_security_module" "test" {
+  name                = "kvHsm%d"
+  resource_group_name = azurerm_resource_group.test.name
+  location            = azurerm_resource_group.test.location
+  sku_name            = "Standard_B1"
+  tenant_id           = data.azurerm_client_config.current.tenant_id
+  admin_object_ids    = [data.azurerm_client_config.current.object_id]
+}
+`, r.template(data), data.RandomInteger)
```

Comment:
rather than inlining this function call - which makes it harder to trace these through, can we split this out into a variable:

```suggestion
`, template, data.RandomInteger)
```